integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/ContainerBasedKafkaCluster.java

```diff
@@ -0,0 +1,226 @@
+
+            kafkaContainer.addFixedExposedPort(holder.getExternalPort(), KAFKA_PORT);
+
+            // Once https://github.com/docker-java/docker-java/pull/1963 is available, we will be able to copy
+            // these properties to the container as server.properties, and avoid the need to map the properties
+            // into environment variables.
+            // kafkaContainer.withCopyFileToContainer(..properties.., "/opt/bitnami/kafka/config/");
+
+            kafkaContainer.withEnv(serverPropertiesToBitnamiEnvVar(holder.getProperties()));
+
+            if (this.clusterConfig.isKraftMode()) {
+                return kafkaContainer
+                        .withEnv("KAFKA_ENABLE_KRAFT", "yes")
+                        .withEnv("KAFKA_KRAFT_CLUSTER_ID", holder.getKafkaKraftClusterId());
+            }
+            else {
+                return kafkaContainer.dependsOn(this.zookeeper);
+            }
+        }).collect(Collectors.toList());
+    }
+
+    private Map<String, String> serverPropertiesToBitnamiEnvVar(Properties serverProperties) {
+        Map<String, String> envVars = new HashMap<>();
+        serverProperties.entrySet().forEach(p -> envVars.put("KAFKA_CFG_" + p.getKey().toString().toUpperCase().replace('.', '_'), p.getValue().toString()));
+        return envVars;
+    }
+
+    @Override
+    public String getBootstrapServers() {
+        return brokers.stream()
+                .map(b -> String.format("localhost:%d", b.getMappedPort(KAFKA_PORT)))
+                .collect(Collectors.joining(","));
+    }
+
+    private Stream<GenericContainer<?>> allContainers() {
+        return Stream.concat(
+                this.brokers.stream(),
+                Stream.ofNullable(this.zookeeper));
+    }
+
+    @Override
+    @SneakyThrows
+    public void start() {
+        try {
+            if (zookeeper != null) {
+                zookeeper.start();
+                Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                    Container.ExecResult result = this.zookeeper.execInContainer(
+                            "sh", "-c",
+                            String.format("_JAVA_OPTIONS=-Dlogback.configurationFile=/tmp/zookeeper_logback.xml zkCli.sh -server zookeeper:%s whoami", ZOOKEEPER_PORT));
+                    LOGGER.debug("zookeeper running {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                    return result.getExitCode() == 0;
+                });
+            }
+            Startables.deepStart(brokers.stream()).get(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException | ExecutionException | TimeoutException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            stop();
+            throw new RuntimeException("startup failed or timed out", e);
+        }
```

Comment:
Can we pull this out to a `startZookeeper` method?

Reply:
The ready check is now performed by the container itself, so I don't think extract method is warrented.